### PR TITLE
CommandOptions

### PR DESCRIPTION
**Pull request by jsmall-nvidia**
Motivation 

* Enable generation of documentation around command line 
* Simplify parsing of command line arguments
* Allow for faster/more robust lookup from values <-> names
* Allow for help information to be targeted  (can specify a category after -h for directed help)
* Keep enumerations up to date in parsing/documentation/display to user
* Remove the *large* and typically out of date help text

The main addition here is CommandOptions structure, that represents "options" and "values". Options being command line options, and "values" akin to enums (although they don't need to be an enum just a list).

Other aspects

* Written in such a way that it isn't tied to *slang* command line parsing
* Uses enum (UserValue) as mechanism to look things/identify. Typically an implementation will use an enum 
* Doesn't *parse*, just simplifies aspects of parsing - uses the same mechanism as before (except with a switch)
* Defining enums to look things up is a little dull, but is straight forward.
* Would allow fairly trivial support for different languages.
* Also fairly trivial to make output markdown for use on website
* Added NameValue types, which simplifies CommandOption setup and other code around such associations
* Usage can specify categories in <>, and doing so will list available options in the description
* Formatting could be improved around lists of things...
* Can generate valid enum lists in diagnostics if category is used for lookup

Future

* Might want make usage machine parsable, at a minimum would be better for extracting flags 
  * Would remove `?...` and `...` from name, which is a little janky
  * Could also be used to check args automagically for some scenarios

```
Usage: slangc [options...] [--] <input files>

General
=======

General options

  -D<name>[=<value>], -D <name>[=<value>]: Insert a preprocessor macro.
  -depfile <path>: Save the source file dependency list in a file.
  -entry <name>: Specify the name of an entry-point function.
    Multiple -entry options may be used in a single invocation. If no -entry 
    options are given, compiler will use [shader(...)] attributes to detect 
    entry points.
  -emit-ir: Emit IR typically as a '.slang-module' when outputting to a 
    container.
  -h or -h <help-category>: Print this message, or help in specified category.
  -I<path>, -I <path>: Add a path to be used in resolving '#include' and 
    'import' operations.
  -lang <language>: Set the language for the following input files.
    <language> can be: c, C, cpp, c++, C++, cxx, slang, glsl, hlsl, cu, cuda
  -matrix-layout-column-major: Set the default matrix layout to column-major.
  -matrix-layout-row-major: Set the default matrix layout to row-major.
  -module-name <name>: Set the module name to use when compiling multiple 
    .slang source files into a single module.
  -o <path>: Specify a path where generated output should be written.
    If no -target or -stage is specified, one may be inferred from file 
    extension (see <file-extension>). If multiple -target options and a single 
    -entry are present, each -o associates with the first -target to its left. 
    Otherwise, if multiple -entry options are present, each -o associates with 
    the first -entry to its left, and with the -target that matches the one 
    inferred from <path>.
  -profile <profile>[+<capability>...]: Specify the shader profile for code 
    generation.
    Accepted profiles are:
      sm_{4_0,4_1,5_0,5_1,6_0,6_1,6_2,6_3,6_4,6_5,6_6}
      glsl_{110,120,130,140,150,330,400,410,420,430,440,450,460}
    Additional profiles that include -stage information:
      {vs,hs,ds,gs,ps}_<version>
    See -capability for information on <capability>
    When multiple -target options are present, each -profile associates with 
    the first -target to its left.
    <capability> can be: spirv_1_{ 0, 1, 2, 3, 4, 5 }, invalid, hlsl, glsl, c, 
    cpp, cuda, spirv_direct, GL_NV_ray_tracing, GL_EXT_ray_tracing, 
    GL_NV_fragment_shader_barycentric, GL_EXT_fragment_shader_barycentric
  -stage <stage>: Specify the stage of an entry-point function.
    When multiple -entry options are present, each -stage associated with the 
    first -entry to its left.
    May be omitted if entry-point function has a [shader(...)] attribute; 
    otherwise required for each -entry option.
    <stage> can be: vertex, hull, domain, geometry, pixel, fragment, compute, 
    raygeneration, intersection, anyhit, closesthit, miss, callable, mesh, 
    amplification
  -target <target>: Specifies the format in which code should be generated.
    <target> can be: unknown, none, hlsl, dxbc, dxbc-asm, dxbc-assembly, dxil, 
    dxil-asm, dxil-assembly, glsl, glsl-vulkan, glsl-vulkan-one-desc, spirv, 
    spirv-asm, spirv-assembly, c, cpp, c++, cxx, torch, torch-binding, 
    torch-cpp, torch-cpp-binding, host-cpp, host-c++, host-cxx, exe, 
    executable, sharedlib, sharedlibrary, dll, cuda, cu, ptx, cuobj, cubin, 
    host-callable, callable, object-code, host-host-callable
  -v, -version: Display the build version.
  -warnings-as-errors all or -warnings-as-errors <id>[,<id>...]: all - Treat 
    all warnings as errors.
    <id>[,<id>...]: Treat specific warning ids as errors.
  -warnings-disable <id>[,<id>...]: Disable specific warning ids.
  -W<id>: Enable a warning with the specified id.
  -Wno-<id>: Disable warning with <id>
  -dump-warning-diagnostics: Dump to output list of warning diagnostic numeric 
    and name ids.
  --: Treat the rest of the command line as input files.

Target
======

Target code generation options

  -capability <capability>[+<capability>...]: Add optional capabilities to a 
    code generation target. See Capabilities below.
    <capability> can be: spirv_1_{ 0, 1, 2, 3, 4, 5 }, invalid, hlsl, glsl, c, 
    cpp, cuda, spirv_direct, GL_NV_ray_tracing, GL_EXT_ray_tracing, 
    GL_NV_fragment_shader_barycentric, GL_EXT_fragment_shader_barycentric
  -default-image-format-unknown: Set the format of R/W images with unspecified 
    format to 'unknown'. Otherwise try to guess the format.
  -disable-dynamic-dispatch: Disables generating dynamic dispatch code.
  -disable-specialization: Disables generics and specialization pass.
  -fp-mode <fp-mode>, -floating-point-mode <fp-mode>: Control floating point 
    optimizations
    <fp-mode> can be: precise, fast, default
  -g, -g<N>, -g<debug-info-format>: Include debug information in the generated 
    code, where possible.
    N is the amount of information, 0..3, unspecified means 2
    <debug-info-format> specifies a debugging info format
    It is valid to have multiple -g options, such as a level and a 
    <debug-info-format>
    <debug-info-format> can be: default-format, c7, pdb, stabs, coff, dwarf
  -line-directive-mode <line-directive-mode>: Sets how the `#line` directives 
    should be produced. Available options are:
    If not specified, default behavior is to use C-style `#line` directives for 
    HLSL and C/C++ output, and traditional GLSL-style `#line` directives for 
    GLSL output.
    <line-directive-mode> can be: none, source-map, default, standard, glsl
  -O<N>: Set the optimization level.
    N is the amount of optimization, 0..3, default is 1
  -obfuscate: Remove all source file information from outputs.

Downstream
==========

Downstream compiler options

  -<compiler>-path <path>: Specify path to a downstream <compiler> executable 
    or library.
    <compiler> can be: none, fxc, dxc, glslang, visualstudio, vs, clang, gcc, 
    genericcpp, c, cpp, nvrtc, llvm
  -default-downstream-compiler <language> <compiler>: Set a default compiler 
    for the given language. See -lang for the list of languages.
    <language> can be: c, C, cpp, c++, C++, cxx, slang, glsl, hlsl, cu, cuda
    <compiler> can be: none, fxc, dxc, glslang, visualstudio, vs, clang, gcc, 
    genericcpp, c, cpp, nvrtc, llvm
  -X<compiler> <option> -X<compiler>... <options> -X.: Pass arguments to 
    downstream <compiler>. Just -X<compiler> passes just the next argument to 
    the downstream compiler. -X<compiler>... options -X. will pass *all* of the 
    options inbetween the opening -X and -X. to the downstream compiler.
    <compiler> can be: none, fxc, dxc, glslang, visualstudio, vs, clang, gcc, 
    genericcpp, c, cpp, nvrtc, llvm
  -pass-through <compiler>: Pass the input through mostly unmodified to the
    existing compiler <compiler>.
    <compiler> can be: none, fxc, dxc, glslang, visualstudio, vs, clang, gcc, 
    genericcpp, c, cpp, nvrtc, llvm

Debugging
=========

Compiler debugging/instrumentation options

  -dump-ast: Dump the AST to a .slang-ast file next to the input.
  -dump-intermediate-prefix <prefix>: File name prefix for -dump-intermediates 
    outputs, default is 'slang-dump-'
  -dump-intermediates: Dump intermediate outputs for debugging.
  -dump-ir: Dump the IR for debugging.
  -dump-ir-ids: Dump the IDs with -dump-ir (debug builds only)
  -dump-repro: Dump a `.slang-repro` file that can be used to reproduce a 
    compilation on another machine.
  -dump-repro-on-error: Dump `.slang-repro` file on any compilation error.
  -E, -output-preprocessor: Output the preprocessing result and exit.
  -extract-repro <name>: Extract the repro files into a folder.
  -load-repro-directory <path>: Use repro along specified path
  -load-repro <name>: Load repro
  -no-codegen: Skip the code generation step, just check the code and generate 
    layout.
  -output-includes: Print the hierarchy of the processed source files.
  -repro-file-system <name>: Use a repro as a file system
  -serial-ir: Serialize the IR between front-end and back-end.
  -skip-codegen: Skip the code generation phase.
  -validate-ir: Validate the IR between the phases.
  -verbose-paths: Display more detailed paths in diagnostic output.
  -verify-debug-serial-ir: Verify IR in the front-end.

Experimental
============

Experimental options (use at your own risk)

  -emit-spirv-directly: Generate SPIR-V output directly (otherwise through GLSL 
    and using the glslang compiler)
  -file-system <fs>: Set the filesystem hook to use for a compile request.
    Accepted file systems: default, load-file, os
  -heterogeneous: Output heterogeneity-related code.
  -no-mangle: Do as little mangling of names as possible.

Internal
========

Internal-use options (use at your own risk)

  -archive-type <archive-type>: Set the archive type for -save-stdlib. Default 
    is zip.
    <archive-type> can be: riff-deflate, riff-lz4, zip, riff
  -compile-stdlib: Compile the StdLib from embedded sources. Will return a 
    failure if there is already a StdLib available.
  -doc: Write documentation for -compile-stdlib
  -ir-compression <type>: Set compression for IR and AST outputs.
    Accepted compression types: none, lite
  -load-stdlib <filename>: Load the StdLib from file.
  -r <name>: reference module <name>
  -save-stdlib <filename>: Save the StdLib modules to an archive file.
  -save-stdlib-bin-source <filename>: Same as -save-stdlib but output the data 
    as a C array.
  -track-liveness: Enable liveness tracking. Places SLANG_LIVE_START, and 
    SLANG_LIVE_END in output source to indicate value liveness.

Depreciated
===========

Deprecated options (allowed but ignored; may be removed in future)

  -parameter-blocks-use-register-spaces: Parameter blocks will use register 
    spaces

<compiler>
==========

Downstream Compilers (aka Pass through)

  none: Unknown
  fxc: FXC HLSL compiler
  dxc: DXC HLSL compiler
  glslang: GLSLANG GLSL compiler
  visualstudio, vs: Visual Studio C/C++ compiler
  clang: Clang C/C++ compiler
  gcc: GCC C/C++ compiler
  genericcpp, c, cpp: A generic C++ compiler (can be any one of visual studio, 
    clang or gcc depending on system and availability)
  nvrtc: NVRTC CUDA compiler
  llvm: LLVM/Clang `slang-llvm`

<language>
==========

Language

  c, C
  cpp, c++, C++, cxx
  slang
  glsl
  hlsl
  cu, cuda

<archive-type>
==============

Archive Type

  riff-deflate
  riff-lz4
  zip
  riff

<line-directive-mode>
=====================

Line Directive Mode

  none: Don't emit `#line` directives at all
  source-map: Use source map to track line associations (doen't emit #line)
  default: Default behavior
  standard: Emit standard C-style `#line` directives.
  glsl: Emit GLSL-style directives with file *number* instead of name.

<debug-info-format>
===================

Debug Info Format

  default-format
  c7
  pdb
  stabs
  coff
  dwarf

<fp-mode>
=========

Floating Point Mode

  precise: Disable optimization that could change the output of floating-point 
    computations, including around infinities, NaNs, denormalized values, and 
    negative zero. Prefer the most precise versions of special functions 
    supported by the target.
  fast: Allow optimizations that may change results of floating-point 
    computations. Prefer the fastest version of special functions supported by 
    the target.
  default: Default floating point mode

<target>
========

Target

  unknown
  none
  hlsl: HLSL source code
  dxbc: DirectX shader bytecode binary
  dxbc-asm, dxbc-assembly: DirectX shader bytecode assembly
  dxil: DirectX Intermediate Language binary
  dxil-asm, dxil-assembly: DirectX Intermediate Language assembly
  glsl: GLSL source code
  glsl-vulkan: GLSL Vulkan source code
  glsl-vulkan-one-desc: GLSL Vulkan source code
  spirv: SPIR-V binary
  spirv-asm, spirv-assembly: SPIR-V assembly
  c: C source code
  cpp, c++, cxx: C++ source code
  torch, torch-binding, torch-cpp, torch-cpp-binding: C++ for pytorch binding
  host-cpp, host-c++, host-cxx: C++ source for host execution
  exe, executable: Executable binary
  sharedlib, sharedlibrary, dll: Shared library/Dll
  cuda, cu: CUDA source code
  ptx: PTX assembly
  cuobj, cubin: CUDA binary
  host-callable, callable: Host callable
  object-code: Object code
  host-host-callable: Host callable for host execution

<stage>
=======

Stage

  vertex
  hull
  domain
  geometry
  pixel, fragment
  compute
  raygeneration
  intersection
  anyhit
  closesthit
  miss
  callable
  mesh
  amplification

<capability>
============

A capability describes an optional feature that a target may or may not 
support. When a -capability is specified, the compiler may assume that the 
target supports that capability, and generate code accordingly.

  spirv_1_{ 0, 1, 2, 3, 4, 5 }: minimum supported SPIR - V version
  invalid
  hlsl
  glsl
  c
  cpp
  cuda
  spirv_direct
  GL_NV_ray_tracing: enables the GL_NV_ray_tracing extension
  GL_EXT_ray_tracing: enables the GL_EXT_ray_tracing extension
  GL_NV_fragment_shader_barycentric: enables the 
    GL_NV_fragment_shader_barycentric extension
  GL_EXT_fragment_shader_barycentric: enables the 
    GL_EXT_fragment_shader_barycentric extension

<file-extension>
================

A <language>, <format>, and/or <stage> may be inferred from the extension of an 
input or -o path

  hlsl, fx: hlsl
  dxbc
  dxbc-asm: dxbc-assembly
  dxil
  dxil-asm: dxil-assembly
  glsl
  vert: glsl (vertex)
  frag: glsl (fragment)
  geom: glsl (geoemtry)
  tesc: glsl (hull)
  tese: glsl (domain)
  comp: glsl (compute)
  slang
  spv: SPIR-V
  spv-asm: SPIR-V assembly
  c
  cpp, c++, cxx: C++
  exe: executable
  dll, so: sharedlibrary/dll
  cu: CUDA
  ptx: PTX
  obj, o: object-code
  zip: container
  slang-module, slang-library: Slang Module/Library
  dir: Container as a directory
```

### CODE REVIEW

File: source/slang/slang-options.cpp [hunk 1]
```diff
@@ -1689,21 +1943,50 @@ struct OptionsParser
...
+                    break;
                 }
-                else if (argValue == "-v" || argValue == "-version")
+                case OptionKind::Version:
                 {
                     sink->diagnoseRaw(Severity::Note, session->getBuildTagString());
+                    break;
                 }
-                else if (argValue == "-h" || argValue == "-help" || argValue == "--help")
+                case OptionKind::Help:
```
Comment (expipiplus1) [inline source/slang/slang-options.cpp:1953]: We should handle `--help` separately, i.e. if any option is `--help` then print the help text and exit(0). This enables one to do `slangc --invalid option --help` and still get something useful. 

See also: https://github.com/shader-slang/slang/issues/2829

**Review by expipiplus1 — APPROVED** [review]
Could we now also be more generous in allowing double dash options, for example `--emit-ir` or `--entry`.  I feel like limiting ourselves to single dash options is largely a hangover from the very early versions of gcc (or predating compilers).

**Comment by expipiplus1** [comment]
This is great!

File: source/slang/slang-diagnostic-defs.h [hunk 2]
```diff
@@ -104,6 +102,9 @@ DIAGNOSTIC(    50, Error, duplicateTargets, "the target '$0' has been specified
 DIAGNOSTIC(    60, Error, cannotDeduceOutputFormatFromPath, "cannot infer an output format from the output path '$0'")
 DIAGNOSTIC(    61, Error, cannotMatchOutputFileToTarget, "no specified '-target' option matches the output path '$0', which implies the '$1' format")
 
+DIAGNOSTIC(    62, Error, unknownCommandLineValue, "unknown value for option. Valid values are '$0'")
```
[the remark this answers is not in the document]
Comment (jsmall-nvidia) [reply source/slang/slang-diagnostic-defs.h:105]: Will add in a subsequent PR.

File: source/slang/slang-options.cpp (hunk 1 above)
Comment (jsmall-nvidia) [reply source/slang/slang-options.cpp:1953]: Will look to fix in later PR.

**Pull request merged**